### PR TITLE
linter: add callStatic check

### PR DESCRIPTION
**Pull request by quasilyte**
Detect when static methods are called using "->" syntax
and instance methods that are called using "::" syntax.

It's not always a mistake, but could be a sign of a code smell.
Also, it makes code harder to read, since it's not apparent
what happens before called method signature is reviewed.

Added "Static" field to meta.FuncInfo to make this check possible.

Signed-off-by: Iskander Sharipov <quasilyte@gmail.com>

### CODE REVIEW

**Comment by quasilyte** [comment]
@YuriyNasretdinov, you're right. Thanks. :)

File: src/linter/block.go [hunk 1]
```diff
@@ -852,8 +857,14 @@ func (b *BlockWalker) handleStaticCall(e *expr.StaticCall) bool {
...
 	e.Call.Walk(b)
 
-	if !ok && !haveMagicMethod(className, `__callStatic`) && !b.r.st.IsTrait {
+	magic := haveMagicMethod(className, `__callStatic`)
+	if !ok && !magic && !b.r.st.IsTrait {
 		b.r.Report(e.Call, LevelError, "undefined", "Call to undefined method %s::%s()", className, methodName)
+	} else {
+		// Method is defined.
+		if !fn.Static && !magic {
+			b.r.Report(e.Call, LevelWarning, "callStatic", "Calling instance method as static method")
```
[the remark this answers is not in the document]
Comment (quasilyte) [reply src/linter/block.go]: I'll try again. Thank you for guidance. :+1:

**Comment by YuriyNasretdinov** [comment]
Any progress on that :)? I think this check is useful, you just need to handle cases like "parent::" calls and maybe "static::" calls, although I think that the only not-really-static call is "parent::"

**Comment by quasilyte** [comment]
I was carried away with other things lately. :(

But I'm planning to add this check at some point.
Maybe not from this PR though, since it can become even more outdated after new parser is merged into master.

**Comment by YuriyNasretdinov** [comment]
Yeah, this is not a lot of code to update though, so I think that should be pretty easy :).

**Comment by quasilyte** [comment]
Closing in favor of #174.

**Pull request closed without merging**